daemon/graphdriver/devmapper/devmapper_test.go

```diff
@@ -40,3 +44,54 @@ func TestDevmapperCreateSnap(t *testing.T) {
 func TestDevmapperTeardown(t *testing.T) {
 	graphtest.PutDriver(t)
 }
+
+func TestDevmapperPoolSizeWithReducedLoopSize(t *testing.T) {
+	driver := reflectGetDeviceMapperDriver(reflectGetGraphDriver(graphtest.GetDriver(t, "devicemapper").(*graphtest.Driver)))
+	defer graphtest.PutDriver(t)
+	// make sure data or metadata loopback size are the default size
+	if s := driver.DeviceSet.Status(); s.Data.Total != uint64(defaultDataLoopbackSize) || s.Metadata.Total != uint64(defaultMetaDataLoopbackSize) {
+		t.Fatalf("data or metadata loop back size is incorrect")
+	}
+	if err := driver.Cleanup(); err != nil {
+		t.Fatal(err)
+	}
+
+	root := strings.TrimRight(driver.home, "/devicemapper")
+	// try shrink data or metadata loopback size
+	tenMB := int64(10 * 1024 * 1024)
+	d, err := graphdriver.GetDriver("devicemapper", root, []string{
+		fmt.Sprintf("dm.loopdatasize=%d", defaultDataLoopbackSize-tenMB),
+		fmt.Sprintf("dm.loopmetadatasize=%d", defaultMetaDataLoopbackSize-tenMB),
+	})
+	if err != nil {
+		t.Fatalf("error creating devicemapper driver: %v", err)
+	}
+	driver = reflectGetDeviceMapperDriver(d)
+	if s := driver.DeviceSet.Status(); s.Data.Total != uint64(defaultDataLoopbackSize) || s.Metadata.Total != uint64(defaultMetaDataLoopbackSize) {
+		t.Fatalf("data or metadata loop back size should not be changed")
+	}
+	if err := driver.Cleanup(); err != nil {
+		t.Fatal(err)
+	}
+
+	// try increase data or metadata loopback size
+	d, err = graphdriver.GetDriver("devicemapper", root, []string{
+		fmt.Sprintf("dm.loopdatasize=%d", defaultDataLoopbackSize+tenMB),
+		fmt.Sprintf("dm.loopmetadatasize=%d", defaultMetaDataLoopbackSize+tenMB),
+	})
+	if err != nil {
+		t.Fatalf("error creating devicemapper driver: %v", err)
+	}
+	driver = reflectGetDeviceMapperDriver(d)
```

Comment:
@cpuguy83 , I failed to get devmapper driver directly from `graphdriver.Driver` since `devmapper.Init` returns unexported object of `naiveDiffDriver `. I changed back to reflect ways.
Also the failed test `TestCleanupMountsAfterCrash` is not related.